### PR TITLE
Cumulate all allow rules with assertions

### LINKED ISSUE #12: BC Break: Since 2.7.2, assertions no longer behave as documented
### BC Break Report

|    Q        |   A
|------------ | ------
| Version     | 2.7.2

#### Summary

According to the documentation, an assertion determines _whether a rule applies_. It doesn't determine whether the rule is an allow rule or a deny rule.

From https://docs.laminas.dev/laminas-permissions-acl/advanced/:

> ...in all other cases (i.e., where a specific role, resource, or privilege is specified for the rule), a failed assertion **results in the rule not applying, and other rules would be used to determine whether access is allowed or denied**.

#### Previous behavior

If a specific 'allow' rule has an assertion that returns false, the rule is skipped, and any other more general rules are applied.

#### Current behavior

If a specific 'allow' rule has an assertion that returns false, it is being treated like a 'deny' rule.

#### How to reproduce

```PHP
use Laminas\Permissions\Acl\Acl;
use Laminas\Permissions\Acl\Assertion\CallbackAssertion;

$acl = new Acl();
$acl->addRole('user');
$acl->addResource('asset');

$acl->allow('user', 'asset', 'read', new CallbackAssertion(
    fn ($acl, $role, $res, $priv) => false
));

$acl->allow('user', null, null);

var_dump($acl->isAllowed('user', 'asset', 'read'));
```

This code prints 'true' with version 2.7.1, but 'false' with version 2.7.2.

### PR DESCRIPTION
**Pull request by PowerKiKi**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This fixes the regression introduced in 624567fe376a70e0bfb5aa8217d5afa13b9d6e61 (2.7.2),
where rules with assertions were no longer cumulated, but instead the process was
aborted on first failed assertion and the privilege was incorrectly denied.

That regression directly contradicted the official documentation that states:

> a failed assertion results in the rule not applying, and other
> rules would be used to determine whether access is allowed or denied.

That regression prevented to fallback on more generic rules. And it also made it very brittle
to work with role inheritance since the order of inheritance declaration became critically important,
while adding/removing privileges anywhere in the inheritance chain could incorrectly have side effect
for other roles in an unexpected way.

So this patch is essentially a revert of the problematic commit. And the use-case at the origin of the
regression is now solved, and tested, differently, with proper usage of (pre-)existing behavior, as it
should have been from the beginning.

So instead of the incorrect:

```php
$acl = new Acl();

$assertAllow = new TestAsset\ChildBooleanAssertion(true);
$assertDeny  = new TestAsset\ChildBooleanAssertion(false);

$acl->addRole('staff');
$acl->addResource('base');
$acl->allow('staff', 'base', 'update', $assertAllow);

$acl->addResource('user', 'base');
$acl->allow('staff', 'user', 'update', $assertDeny);
```

It should be:

```php
$acl = new Acl();
$assertAllow = new MockAssertion(true);

$acl->addRole('staff');
$acl->addResource('base');
$acl->allow('staff', 'base', 'update', $assertAllow);

$acl->addResource('user', 'base');
$acl->deny('staff', 'user', 'update', $assertAllow);
```
Fixes #12

### CODE REVIEW

**Comment by PowerKiKi** [comment]
This unfortunately is a BC break because the regression itself was a BC break. Though I'd say it should still be released as patch version since it mostly is a bugfix...

**Comment by PowerKiKi** [comment]
@boesing, I am aware that you are not super familiar with this code base, but maybe you could still give it a "light" review and/or assign a reviewer to it ?

**Comment by PowerKiKi** [comment]
@froschdesign would you be able to review this yourself, or alternatively recommend another reviewer, so that the PR does not get too old ?

File: src/Acl.php [hunk 1]
```diff
@@ -1024,11 +1024,24 @@ protected function getRuleType(
...
+            );
+        }
+
+        if (null === $rule['assert'] || $assertionValue) {
+            return $rule['type'];
+        } elseif (null !== $resource || null !== $role || null !== $privilege) {
+            return;
+        } elseif (self::TYPE_ALLOW === $rule['type']) {
+            return self::TYPE_DENY;
         }
```
Comment (weierophinney) [inline src/Acl.php]:
```suggestion
        if (null === $rule['assert'] || $assertionValue) {
            return $rule['type'];
        }
        
        if (null !== $resource || null !== $role || null !== $privilege) {
            return;
        }
        
        if (self::TYPE_ALLOW === $rule['type']) {
            return self::TYPE_DENY;
        }
```

There's no need for an `elseif` if you are returning from the block. Additionally, it makes it easier to identify the distinct exit conditions.

File: test/AclTest.php [hunk 2]
```diff
@@ -1466,18 +1466,50 @@ public function testAllowNullPermissionAfterResourcesExistShouldAllowAllPermissi
     /**
      * @see https://github.com/laminas/laminas-permissions-acl/issues/2
      */
-    public function testWhenAssertionReturnsFalseTheInversionOfItsTypeShouldBeUsed()
+    public function testCanDenyAnInheritedRuleThatAllows()
```
Comment (weierophinney) [inline test/AclTest.php]:
```suggestion
    public function testCanDenyAnInheritedAllowRule()
```

As a native English speaker, I read this and thought "allows what?"; the suggested change makes it more clear that you are denying a previously set rule.

**Review by weierophinney — DISMISSED** [review]
Overall, this looks good; I've just provided some suggestions to tighten it up.

**Comment by PowerKiKi** [comment]
Thank you for the review. I applied all your suggestions. IMHO this is now ready to be merged.

**Comment by PowerKiKi** [comment]
@snapshotpl you attempted to rebase this PR a few hours ago. I did it for you. And since it has been reviewed by @weierophinney already, could you please considering merging it ?

**Comment by froschdesign** [comment]
@PowerKiKi 
Your pull request will be released with [version 2.9.0](https://github.com/laminas/laminas-permissions-acl/milestone/6).

**Comment by Ocramius** [comment]
Removing the `BC Break` label, since:

 1. we decided this is suited for a minor release
 2. it's actually fixing a BC Break (label would be confusing in release notes)

**Review by Ocramius — APPROVED** [review]
:ship:

**Pull request merged**